src/optimizer/rule_impls.cpp

```diff
@@ -1004,5 +1106,104 @@ void PullFilterThroughMarkJoin::Transform(
   transformed.push_back(output);
 }
 
+///////////////////////////////////////////////////////////////////////////////
+/// PullFilterThroughAggregation
+PullFilterThroughAggregation::PullFilterThroughAggregation() {
+  type_ = RuleType::PULL_FILTER_THROUGH_AGGREGATION;
+
+  auto filter = std::make_shared<Pattern>(OpType::LogicalFilter);
+  filter->AddChild(std::make_shared<Pattern>(OpType::Leaf));
+  match_pattern = std::make_shared<Pattern>(OpType::LogicalAggregateAndGroupBy);
+  match_pattern->AddChild(filter);
+}
+
+int PullFilterThroughAggregation::Promise(GroupExpression *group_expr,
+                                          OptimizeContext *context) const {
+  (void)context;
+  auto root_type = match_pattern->Type();
+  // This rule is not applicable
+  if (root_type != OpType::Leaf && root_type != group_expr->Op().type()) {
+    return 0;
+  }
+  return static_cast<int>(UnnestPromise::High);
+}
+
+bool PullFilterThroughAggregation::Check(
+    std::shared_ptr<OperatorExpression> plan, OptimizeContext *context) const {
+  (void)context;
+  (void)plan;
+
+  auto &children = plan->Children();
+  PL_ASSERT(children.size() == 1);
+  UNUSED_ATTRIBUTE auto &r_grandchildren = children[1]->Children();
+  PL_ASSERT(r_grandchildren.size() == 1);
+
+  return true;
+}
+
+void PullFilterThroughAggregation::Transform(
+    std::shared_ptr<OperatorExpression> input,
+    std::vector<std::shared_ptr<OperatorExpression>> &transformed,
+    UNUSED_ATTRIBUTE OptimizeContext *context) const {
+  LOG_TRACE("PullFilterThroughAggregation::Transform");
+  auto &memo = context->metadata->memo;
+  auto &filter_expr = input->Children()[0];
+  auto child_group_id =
+      filter_expr->Children()[0]->Op().As<LeafOperator>()->origin_group;
+  const auto &child_group_aliases_set =
+      memo.GetGroupByID(child_group_id)->GetTableAliases();
+
+  auto &predicates = filter_expr->Op().As<LogicalFilter>()->predicates;
+
+  std::vector<AnnotatedExpression> correlated_predicates;
+  std::vector<AnnotatedExpression> normal_predicates;
+  std::vector<std::shared_ptr<expression::AbstractExpression>> new_groupby_cols;
+  for (auto &predicate : predicates) {
+    if (util::IsSubset(child_group_aliases_set, predicate.table_alias_set)) {
+      normal_predicates.emplace_back(predicate);
+    } else {
+      // Correlated predicate, already in the form of
+      // (outer_relation.a = (expr))
```

Comment:
Where do we check or enforce the correlated predicate to be in the form of `outer_relation.a = (expr)`? If the predicate is in the form of inequality (e.g. out_relation.a > inner_relation.b), the result can be incorrect.

Reply:
I've check these in `QueryToOperatorTransformer::IsSupportedSubSelect`. I added two checkers to tell if the subselect/correlated expression is valid before transforming to operator tree, I think it should be clearer this way.

Reply:
I see. Then I think it should be fine.